pkg/estimator/server/estimate.go

```diff
@@ -110,3 +111,100 @@ func (es *AccurateSchedulerEstimatorServer) nodeMaxAvailableReplica(node *framew
 }
+
+// SelectVictims returns max available replicas in terms of request and cluster status.
+func (es *AccurateSchedulerEstimatorServer) SelectVictims(ctx context.Context, request *pb.PreemptionRequest) ([]pb.ObjectReference, error) {
+	trace := utiltrace.New("Preempting", utiltrace.Field{Key: "namespacedName", Value: request.Preemptor})
+	defer trace.LogIfLong(100 * time.Millisecond)
+
+	snapShot := schedcache.NewEmptySnapshot()
+	if err := es.Cache.UpdateSnapshot(snapShot); err != nil {
+		return []pb.ObjectReference{}, err
+	}
+	trace.Step("Snapshotting estimator cache and node infos for preemption completed")
+
+	if snapShot.NumNodes() == 0 {
+		return []pb.ObjectReference{}, nil
+	}
+
+	candidateVictims, err := es.selectVictims(ctx, snapShot, request)
+	if err != nil {
+		return []pb.ObjectReference{}, err
+	}
+	trace.Step("Preemption completed")
+
+	return *candidateVictims, nil
+}
+
+func (es *AccurateSchedulerEstimatorServer) selectVictims(
+	ctx context.Context,
+	snapshot *schedcache.Snapshot,
+	request *pb.PreemptionRequest,
+) (*[]pb.ObjectReference, error) {
+	allNodes, err := snapshot.NodeInfos().List()
+	if err != nil {
+		return &[]pb.ObjectReference{}, err
+	}
+	var (
+		requirements     = request.ReplicaRequirements
+		requiredReplicas = request.NumReplicas
+		affinity         = nodeutil.GetRequiredNodeAffinity(requirements)
+		tolerations      []corev1.Toleration
+	)
+
+	if requirements.NodeClaim != nil {
+		tolerations = requirements.NodeClaim.Tolerations
+	}
+	candidateVictims, err := preemptee.ListCandidateVictims(request.Preemptor.Kind, &es.PreempteeLister)
+	if err != nil {
+		return &[]pb.ObjectReference{}, err
+	}
+	// run preemption plugins to filter out eligible victims for preemption
+	candidateVictims, ret := es.estimateFramework.RunPreemptionFilterPlugins(ctx, candidateVictims, request)
+
+	// No victims can be preempted on the cluster, skip further checks and return 0
+	if ret.IsUnschedulable() {
+		return &[]pb.ObjectReference{}, nil
+	}
+
+	if !ret.IsSuccess() && !ret.IsNoOperation() {
+		return &[]pb.ObjectReference{}, fmt.Errorf(fmt.Sprintf("preemption filter plugins fails with %s", ret.Reasons()))
+	}
+
+	orderedCandidateVictims, ret := es.estimateFramework.RunPreemptionOrderPlugins(ctx, candidateVictims, request)
+	if !ret.IsSuccess() && !ret.IsNoOperation() {
+		return &[]pb.ObjectReference{}, fmt.Errorf(fmt.Sprintf("preemption order plugins fails with %s", ret.Reasons()))
+	}
+	victimResourceBindings := make([]pb.ObjectReference, len(orderedCandidateVictims))
+	for _, candidateVictim := range orderedCandidateVictims {
+		victimResourceBindings = append(victimResourceBindings, *candidateVictim.ResourceBinding)
+	}
+	for cnt, candidateVictim := range orderedCandidateVictims {
+		var victimRbs int32
+		processNode := func(i int) {
+
+			node := allNodes[i]
+			if !nodeutil.IsNodeAffinityMatched(node.Node(), affinity) || !nodeutil.IsTolerationMatched(node.Node(), tolerations) {
```

Comment:
List returns
```
List(selector labels.Selector) (ret []*v1.Node, err error)
```
node is a pointer, you might need to substract the resource from node.Allocatable after processNode.

Reply:
should I make a clone copy? I think this allNodes is snapshotted at the start of the function and only used inside processNode to validate all the nodes can actually fit enough copies of the preemptor.